### PR TITLE
[TreeView] Add `selectItem` and `getItemDOMElement` methods to the public API

### LINKED ISSUE #10113: [TreeView] Scroll to programmatically selected tree item
### Summary 💡

When selection is brought to a tree item programmatically, that page should scroll to that item in the tree view component.

### Examples 🌈

https://github.com/mui/mui-x/assets/3165635/561c3db6-3079-4799-b62a-61d4fea0453f

### Motivation 🔦

The TreeView component acts as a page hierarchy explorer for Toolpad with content often overflowing beyond its container. With this feature, the tree view component would scroll to the item currently selected in the page (and the tree itself as well)

### Order ID 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by flaviendelangle**
Closes #10113

[selectItem doc preview](https://deploy-preview-13485--material-ui-x.netlify.app/x/react-tree-view/rich-tree-view/selection/#select-or-deselect-an-item)
[getItemDOMElement doc preview](https://deploy-preview-13485--material-ui-x.netlify.app/x/react-tree-view/rich-tree-view/items/#get-an-items-dom-element-by-id)

@bharatkashyap I propose to expose a method that allow you to easily retrieve the DOM element rather than a method that allow you to scroll.
I first tried to implement something like `apiRef.current.scrollToItem(itemId)`, but I was basically copy-pasting the entire API of the `scrollIntoView` method.

Your use case can be achieved as follow:

```ts
apiRef.current!.selectItem({ event, itemId, isSelected: true });
apiRef.current!.getItemDOMElement(itemId)?.scrollIntoView({ block: 'center' });
```

And if you are using multi selection you might prefer:

```ts
apiRef.current!.selectItem({ event, itemId, isSelected: true, keepExistingSelection: true });
apiRef.current!.getItemDOMElement(itemId)?.scrollIntoView({ block: 'center' });
```

_____

I'll add some tests once we agree on the DX

### CODE REVIEW

File: docs/data/tree-view/rich-tree-view/expansion/expansion.md [hunk 1]
```diff
@@ -56,9 +56,10 @@ Use the `setItemExpansion` API method to change the expansion of an item.
 apiRef.current.setItemExpansion(
   // The DOM event that triggered the change
   event,
-  // The ID of the item to expand or collapse
+  // The id of the item to expand or collapse
```
Comment (flaviendelangle) [inline docs/data/tree-view/rich-tree-view/expansion/expansion.md:59]: I tried to unify a bit the wording for those boolean parameters throughout the Tree View codebase
And to always have the same descriptions for the params in the JSDoc and in the doc section

**Comment by bharatkashyap** [comment]
@flaviendelangle This DX looks perfectly agreeable to me 👍🏻

File: packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.ts [hunk 2]
```diff
@@ -88,16 +88,16 @@ export const useTreeViewSelection: TreeViewPlugin<UseTreeViewSelectionSignature>
...
-      } else if (!isSelectedBefore && (newValue === true || newValue == null)) {
+      } else if (!isSelectedBefore && (isSelected === true || isSelected == null)) {
         newSelected = [itemId].concat(cleanSelectedItems);
       } else {
         newSelected = cleanSelectedItems;
       }
     } else {
       // eslint-disable-next-line no-lonely-if
-      if (newValue === false) {
+      if (isSelected === false || (isSelected == null && instance.isItemSelected(itemId))) {
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.ts:100]: Small fix in a scenario that was never used before (single selection, `isSelected` not defined and the item is selected)

File: packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.types.ts [hunk 3]
```diff
@@ -3,52 +3,56 @@ import type { DefaultizedProps, TreeViewPluginSignature } from '../../models';
...
 
-export interface UseTreeViewSelectionInstance {
+export interface UseTreeViewSelectionPublicAPI {
+  /**
+   * Select or deselect an item.
+   * @param {object} params The params of the method.
+   * @param {React.SyntheticEvent} params.event The DOM event that triggered the change.
+   * @param {string} params.itemId The id of the item to select or deselect.
+   * @param {boolean} params.keepExistingSelection If `true`, the other already selected items will remain selected, otherwise, they will be deselected. This parameter is only relevant when `multiSelect` is `true`
+   * @param {boolean | undefined} params.isSelected If `true` the item will be selected. If `false` the item will be deselected. If not defined, the item's new selection status will be the opposite of its current one.
```
Comment (noraleonte) [inline packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.types.ts:13]: The naming of `isSelected` is slightly confusing to me. I would expect something that begins with `is[...]` to define a state, but here isSelected defines what should **happen** to the item (if it should be selected or not). Is there a drawback to renaming this to something like `selectItem`?

Comment (flaviendelangle) [reply packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.types.ts:13]: `shouldBeSelected`? To follow a boolean naming :angel:

Comment (noraleonte) [reply packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.types.ts:13]: sounds good to me 👌

**Comment by flaviendelangle** [comment]
> expandedItems is the most powerful since it is "controlled", thus using selectItem method cannot possibly internally change what is "controlled" externally. This aught to be mentioned in the docs.

If you provide `expandedItems`, then you should also provide a `onExpandedItemsChange` to update your model whenever it changes.
We could add a note on the doc.

> I suggest to add a new method to apiRef which returns an array of nodes' ids which leads the path all the way down to the selectItem (supplied nodeId). This result could be manually used when expandedItems is used (as a state controlled externally)

Could you expand on this?

**Comment by flaviendelangle** [comment]
> I had no idea that onExpandedItemsChange is automatically called if apiRef.current.selectItem() method is called. That should be documented.

Well, if you call `selectItem`, you change the selected items, so `onExpandedItemsChange` is called just like for any other expansion change.

> I assumed, logically, that TreeView will never change a controlled state "under the hood"

If you control `expandedItems` and your user expands an item on the UI, it wil call `onExpandedItemsChange` to update the controlled model (and you can refuse to update it in which case the UI will not expand the element).
The behavior is exactly the same with `apiRef.selectItem`, if you control the model, you can choose to update it or not.

But note that the goal of those imperative methods is to allow you to update imperatively the expanded items without having to control the model to do it.
You can still control it if you have other needs that requires it, but if you are only controlling it to expand imperatively an item, then you should be able to not control it at all.

For your code snippet, if I understand correctly your use case, you are trying to expand a deeply nested item while making sure all it's parent are also expanded in the process (to always have the selected item always visible).
The way to do it in our new API would  probably either:

1. To add a new `shouldExpandAncestors` property to the `apiRef.selectItem()` method. If `true`, selecting an item would automatically expand all its collapsed ancestors to make sure the item is visible.

  ```ts
  apiRef.current.selectItem({
    event,
    itemId,
    shouldExpandAncestors: true,
  });
  ```

2. To add a new `shouldExpandAncestors` property to the `apiRef.setItemExpansion()` method. This is slightly less magic and you have to do the two imperative calls yourselve. It also allow to react in a `useEffect` if you are not using `apiRef.selectItem` to select the items but instead directly updating your model. But it would require to rework the params of `setItemExpansion` (which is planned regardless):

  ```ts
  apiRef.current.setItemExpansion({
    event,
    itemId,
    isExpanded: true,
    shouldExpandAncestors: true,
  });
  ```


___

I think it's better to have this kind of behavior built in inside the component because it's way easier to implement for the end user and the use case is far from being niche.
But, if you have some specific requirement that prevent you from using those built-in imperative methods and you need to keep controlling both the expansion and the selection, then you indeed need a way to build the path leading to an item to then update your model.
I think it would make sense to expose a `getItemParentId` method (or `getItemMeta` but I'm not a fan of making the whole `TreeViewItemMeta` method pubic. But the exact DX of this new methods could be discussed.

**Comment by yairEO** [comment]
what if I want to use `setItemExpansion` as a toggler without explicitly setting `isExpanded: true, shouldExpandAncestors: true,` but toggle these, on click somewhere, without caring what is the current state (expanded or collapsed).

> if I understand correctly your use case, you are trying to expand a deeply nested item while making sure all it's parent are also expanded

For that I wrote a utility method which traverses the tree upwards from a certain given `node id` and I use that for the `expandedItems` prop.

I have a complex tree where I need need to programmatically expand and collapse multiple branches in the tree, to visually show only the portions of the tree which are relevant to certain different scenarios in my app.

`setItemExpansion` is singular and so cannot be easily used to toggle multiple nodes at once (and their ancestors)

Therefore for me working with a controlled `expanded` state gives me maximum freedom, with some utility methods which I've written. 

For example, I've added a prop for my Tree wrapper for `initialExpansionLevel` and a utility method is scanning the tree, generating an array of expanded tree nodes ids for a given desired level. Very handy for complex tree based on a lot of data where you want to show 10-30 items but not hundreds, and you do not care about the names, only how many levels to initially expand.

**Pull request merged**